google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc_test.cc

```diff
@@ -0,0 +1,462 @@
+            retry_policy_factory_.AsStdFunction(), backoff_policy_,
+
+            sleeper_.AsStdFunction(), stream_factory_.AsStdFunction(),
+            initializer_.AsStdFunction())} {}
+
+  StrictMock<MockFunction<future<void>(std::chrono::milliseconds)>> sleeper_;
+  StrictMock<MockFunction<AsyncReadWriteStreamReturnType()>> stream_factory_;
+  StrictMock<MockFunction<std::unique_ptr<MockRetryPolicy>()>>
+      retry_policy_factory_;
+  StrictMock<MockFunction<future<StatusOr<AsyncReadWriteStreamReturnType>>(
+      AsyncReadWriteStreamReturnType)>>
+      initializer_;
+  std::shared_ptr<StrictMock<MockBackoffPolicy>> backoff_policy_;
+  ResumableAsyncReadWriteStream stream_;
+};
+
+TEST_F(ResumableAsyncReadWriteStreamingRpcTest, SingleStartFailureThenGood) {
+  InSequence seq;
+
+  auto stream1 = absl::make_unique<StrictMock<AsyncReaderWriter>>();
+  auto& stream1_ref = *stream1;
+
+  auto mock_retry_policy = absl::make_unique<StrictMock<MockRetryPolicy>>();
+  auto& retry_policy_ref = *mock_retry_policy;
+  EXPECT_CALL(retry_policy_factory_, Call)
+      .WillOnce(Return(ByMove(std::move(mock_retry_policy))));
+
+  auto backoff_policy = absl::make_unique<StrictMock<MockBackoffPolicy>>();
+  auto& backoff_policy_ref = *backoff_policy;
+  EXPECT_CALL(*backoff_policy_, clone)
+      .WillOnce(Return(ByMove(std::move(backoff_policy))));
+
+  EXPECT_CALL(stream_factory_, Call)
+      .WillOnce(Return(ByMove(std::move(stream1))));
+
+  promise<bool> start_promise;
+  EXPECT_CALL(stream1_ref, Start)
+      .WillOnce(Return(ByMove(start_promise.get_future())));
+
+  auto status_future = stream_->Start();
+
+  EXPECT_CALL(stream1_ref, Finish)
+      .WillOnce(Return(ByMove(make_ready_future(kFailStatus))));
+  EXPECT_CALL(retry_policy_ref, IsExhausted).WillOnce(Return(false));
+  EXPECT_CALL(retry_policy_ref, OnFailure(kFailStatus)).WillOnce(Return(true));
+  EXPECT_CALL(backoff_policy_ref, OnCompletion)
+      .WillOnce(Return(std::chrono::milliseconds(7)));
+  EXPECT_CALL(sleeper_, Call).WillOnce(ZeroSleep());
+
+  auto stream2 = absl::make_unique<StrictMock<AsyncReaderWriter>>();
+  auto& stream2_ref = *stream2;
+  EXPECT_CALL(stream_factory_, Call)
+      .WillOnce(Return(ByMove(std::move(stream2))));
+  EXPECT_CALL(stream2_ref, Start)
+      .WillOnce(Return(ByMove(make_ready_future(true))));
+
+  EXPECT_CALL(initializer_, Call).WillOnce(InitializeInline());
+
+  start_promise.set_value(false);
+
+  // start future doesn't finish until
+  // permanent error or `Finish`is called
+  EXPECT_EQ(status_future.wait_for(ms(kFutureWaitMs)),
+            std::future_status::timeout);
+
+  EXPECT_CALL(stream2_ref, Finish)
+      .WillOnce(Return(ByMove(make_ready_future(Status()))));
+  auto shutdown = stream_->Shutdown();
+  shutdown.get();
+  EXPECT_THAT(status_future.get(), IsOk());
+}
+
+TEST_F(ResumableAsyncReadWriteStreamingRpcTest,
+       SingleStartFailurePermanentError) {
```

Comment:
I am missing a test for "too many transient failures during start".

Reply:
This will, in practice, never happen, as the streams should retry forever on transient errors like network outages. But we can add a test that this fails permanently when IsExhausted returns true

Reply:
Hmmm... If I understand what you are suggesting then if the customer misconfigures the library (for example, but not limited to: set the endpoint to something that rejects connections) they would get no notification whatsoever, the library will just keep retrying after getting `kUnavailable`.

In other libraries we set a limit (time or error count base) on retries for unsuccessful streams (i.e., those that get no data ever).  If the stream gets some data we allow then retry regardless of the error code.

Reply:
added a test in which a permanent error occurred via `IsExhausted` returning `true`